tests/bsim/bluetooth/host/misc/hfc/src/main.c

```diff
@@ -0,0 +1,444 @@
+}
+
+static bool is_disconnected(struct bt_conn *conn)
+{
+	int err;
+	struct bt_conn_info info;
+
+	err = bt_conn_get_info(conn, &info);
+	ASSERT(err == 0, "Failed to get info for %p\n", conn);
+
+	/* Return if fully disconnected */
+	return info.state == BT_CONN_STATE_DISCONNECTED;
+}
+
+static int disconnect(struct bt_conn *conn)
+{
+	if (!conn) {
+		return -ENOTCONN;
+	}
+
+	int err = bt_conn_disconnect(conn, BT_HCI_ERR_REMOTE_USER_TERM_CONN);
+
+	if (!err) {
+		WAIT_FOR_FLAG_UNSET(is_connected);
+	}
+
+	return err;
+}
+
+void entrypoint_dut(void)
+{
+	/* Test purpose:
+	 *
+	 * Verifies that there is no host RX buffer leak due to disconnections.
+	 *
+	 * That is, not actual host buffers (ie memory) but rather the number of
+	 * free buffers that the controller thinks the host has.
+	 *
+	 * If there is a desynchronization between those two, the result is that
+	 * the controller stops forwarding ACL data to the host, leading to an
+	 * eventual application timeout.
+	 *
+	 * To do this, the DUT is connected to a peer that loops through sending
+	 * a few ATT notifications then disconnecting.
+	 *
+	 * The test stops after an arbitrary number of notifications have been
+	 * received.
+	 *
+	 * [verdict]
+	 * - no buffer allocation failures, timeouts or stalls.
+	 */
+	int err;
+	struct dut_state *s = &g_dut_state;
+
+	LOG_DBG("Test start: DUT");
+
+	s->rx = 0;
+
+	err = bt_enable(NULL);
+	ASSERT(err == 0, "Can't enable Bluetooth (err %d)\n", err);
+	LOG_DBG("Central: Bluetooth initialized.");
+
+	s->conn = connect_and_subscribe();
+
+	LOG_DBG("Central: Connected and subscribed to both peers");
+
+	/* Wait until we got all notifications from both peers */
+	while (s->rx < NUM_NOTIFICATIONS) {
+		LOG_DBG("%d packets left, waiting..",
+			NUM_NOTIFICATIONS - s->rx);
+		k_msleep(100);
+		if ((s->rx < NUM_NOTIFICATIONS) && is_disconnected(s->conn)) {
+			LOG_INF("reconnecting..");
+			/* release the ref we took in the `connected` callback */
+			bt_conn_unref(s->conn);
+
+			/* release the ref we took when starting the scanner */
+			bt_conn_unref(s->conn);
+
+			s->conn = NULL;
+			s->conn = connect_and_subscribe();
+		}
+	}
+
+	/* linux will "unref" the conn :p */
+	disconnect(s->conn);
+
+	PASS("DUT done\n");
+}
+
+void entrypoint_peer_0(void)
+{
+	int err;
+	int tx;
+	struct bt_conn *conn;
+	const struct bt_gatt_attr *attr;
+	uint8_t data[10];
+
+	LOG_DBG("Test start: peer 0");
+
+	err = bt_enable(NULL);
+	ASSERT(err == 0, "Can't enable Bluetooth (err %d)\n", err);
+	LOG_DBG("Bluetooth initialized.");
+
+	/* prepare data for notifications */
+	attr = &test_gatt_service.attrs[2];
+	memset(data, 0xfe, sizeof(data));
+
+	/* Pass unless something else errors out later */
+	PASS("peer 0 done\n");
+
+	tx = 0;
+	while (true) {
+		conn = connect_as_peripheral();
+
+		LOG_INF("wait until DUT subscribes");
+		UNSET_FLAG(is_subscribed);
+		WAIT_FOR_FLAG(is_subscribed);
+
+		LOG_INF("send notifications");
+		for (int ltx = 0; ltx < 10; ltx++) {
```

Comment:
What is `ltx` short for?

Reply:
"local" tx. It was because before, this was not a while-true. There were two for-loops with TX batches.
I'll rename it to the more meaningful `i` :p